articles/avoid-fix-flaky-tests.md

```diff
@@ -0,0 +1,73 @@
+
+
+# Analyze Failures Systematically
+
+## Reproduce Failures Locally
+
+If you have to rely on a CI server and test environment to reproduce a failure, your feedback loop after every change will be very long. This makes fixing the issue a cumbersome and inefficient procedure. Therefore make sure you can start the test locally with the debugger attached. If the test does not fail as often as on the CI, consider running the test in a loop to get a local failure quickly.
+
+Some failures will not be reproducable locally though and you will be forced to analyze them on the CI environment. One of the build nodes could temporarily serve for the analysis of the issue while the others can still be used by the rest of the team.
+
+## Collect Data From Many Runs
+
+If you have to deal with flaky tests that fail very rarely, running the tests a few times a day might not give you enough data for analysis. Therefore schedule the tests to run as often as you can afford. In a recent project we ran the test suite once every hour throughout the night. This gave us a lot of data to analyze stability the next morning.
+
+## Aggregate Results From Feature Branches
+
+You might actually have a lot of data from running the tests on each feature branch, but it is hard to see patterns if you don't have the test runs in an aggregated form. Collect all the test failures from all branches in a single place to analyze them. You might start off with an Excel sheet, but will probably automate the data collection soon.
+
+## Visualize
+
+Visualize failures in a matrix where one dimension is the test run and the other is the test scenario (like the [Jenkins Test Results Analyzer Plugin](https://wiki.jenkins.io/display/JENKINS/Test+Results+Analyzer+Plugin) does. This might reveal patterns that can give you a hint where the problem lies. E.g. if a certain group of tests always fail together, you might want to check what they have in common. And if a single test only fails every 10th time or so, painting it on a graph will make it apparent that there's a problem with this test.
+
+## Analyze Logs and Metrics
+
+Logs and metrics from the CI server that runs the test and from the SUT might give you valuable hints to the root cause. If your servers run in VMs, you might also want to collect metrics and log files from the VM host. Collect everything in a central place so you can analyze the timeline of events and look for reoccurring patterns.
+
+
+# If All Else Fails
+
+## Rerun Failed Tests Automatically
+
+If you can't tame the beast or if you need a quick fix to buy you some time for a proper solution you might want to rerun failed tests once or twice automatically (in Java you could use a JUnit rule for this). If the reruns is successful, the build is successful. This will make your build runs a bit slower, but might spare you a rerun of the entire build pipeline.
```

Comment:
Always beware of the broken window effect tho. ;-)

Reply:
Added sentence: This approach might have the undesirable side effect that people care less about writing stable tests as flaky ones don't hurt so much anymore.